### PR TITLE
Complex Entity for AOI

### PR DESCRIPTION
**Pull request by sayas01**
### Description:

This PR adds new feature to Atlas to include Complex Entity support for Areas Of Interest (AOI). This PR addresses the OSM issue [#163](https://github.com/osmlab/atlas/issues/163). Two new classes added are:
1. ComplexAOI : ComplexEntity implementation of Multipolygon AOI relations and areas. 
2. ComplexAOIFinder : For the given Atlas, builds complex entities for all the AOIs that are represented by Multipolygon relations and areas.
Currently, the implementation uses a small subset of 600+ key-value pairs of AOI tags (filtered from https://wiki.openstreetmap.org/wiki/Map_Features). The source of the subset AOI tags is the [OverlappingAOIPolygonCheck](https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java). 

### Potential Impact:

The new feature could be made use of in getting geometries of AOI.

### Unit Test Approach:

Added test cases for the ComplexAOIFinder class.

### Test Results:

All tests passed.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+     * meets the requirements for Complex AOI relation. The AOI tags are checked against the default
+     * tags.
+     *
+     * @param source
+     *            The {@link AtlasEntity} for which the ComplexEntity is created
+     * @return {@link ComplexAOIRelation} if created, else return empty.
+     */
+    public static Optional<ComplexAOIRelation> getComplexAOIRelation(final AtlasEntity source)
+    {
+        return !(source instanceof Relation && AOI_TAG_FILTER.test(source)) ? Optional.empty()
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java]: Could we switch these around? So
```
(source instanceof Relation && AOI_TAG_FILTER.test(source)) ? Optional.of(...) : Optional.empty();
```

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     * @param aoiFilter
+     *            The {@link TaggableFilter} of AOI tags against which the relation is checked for
+     *            AOI tags
+     * @return {@link ComplexAOIRelation} if created, else return empty.
+     */
+    public static Optional<ComplexAOIRelation> getComplexAOIRelation(final AtlasEntity source,
+            final TaggableFilter aoiFilter)
+    {
+        return !(source instanceof Relation
+                && (AOI_TAG_FILTER.test(source) || aoiFilter.test(source))) ? Optional.empty()
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java]: See above. Also, if there's a `TaggableFilter` that matches nothing, we could probably have the last function call this one 😄

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        super(source);
+        try
+        {
+            this.multiPolygon = RELATION_TO_MULTI_POLYGON_CONVERTER.convert(source);
+        }
+        catch (final Exception exception)
+        {
+            logger.warn("Unable to create complex relations from {}", source, exception);
+            setInvalidReason("Unable to create complex islands", exception);
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java]: nit: complex AOIs?

**Review by nachtm — CHANGES_REQUESTED** [review]
🔥 This is awesome! A couple style comments, but this is a nicely-done ComplexEntity 🙂

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * to the default {@link TaggableFilter} of AOI tags if the tags are not explicitly specified.
+ *
+ * @author sayas01
+ */
+public final class ComplexAOIRelation extends ComplexEntity
+{
+    private static final Logger logger = LoggerFactory.getLogger(ComplexAOIRelation.class);
+    private static final RelationOrAreaToMultiPolygonConverter RELATION_TO_MULTI_POLYGON_CONVERTER = new RelationOrAreaToMultiPolygonConverter();
+    // The default AreasOfInterest(AOI) tags
+    private static final TaggableFilter AOI_TAG_FILTER = TaggableFilter.forDefinition(
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/ComplexAOIRelation.java]: I would consider turning this into a json resource, as I can see the list of aois growing substantially.

**Review by nachtm — APPROVED** [review]
🎉 ❗️

**Review by Bentleysb — APPROVED** [review]
LGTM

File: src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/aoi-tag-filter.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{
+  "filters": [
+    "aerialway->STATION",
```
Comment (danielduhh) [inline src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/aoi/aoi-tag-filter.json:3]: Nice! Lets make sure to mention that the OSM wiki was the source for this

**Review by danielduhh — APPROVED** [review]
Looks good! Thanks @sayas01

**Pull request closed without merging**